### PR TITLE
fix #13899 defer now works with async

### LINKED ISSUE #13899: Defer is not properly working for asynchronous procedures.
It looks like `defer` has some bug.

### Example
```nim
import asyncdispatch

proc someConnect() {.async.} =
  await sleepAsync(100)

proc someClose() {.async.} =
  await sleepAsync(1000)

proc testFooFails(): Future[bool] {.async.} =
  await someConnect()
  defer:
    await someClose()
    result = true

proc testFooSucceed(): Future[bool] {.async.} =
  try:
    await someConnect()
  finally:
    await someClose()
    result = true

when isMainModule:
  echo waitFor testFooFails()
  echo waitFor testFooSucceed()
```

### Current Output
```
false
true
```

### Expected Output
```
true
true
```

### Additional Information
* At least 1.0.4, 1.0.6 and 1.2.0 are affected.

### PR DESCRIPTION
**Pull request by timotheecour**
fix #13899

CI failure unrelated, and I have a fix for it https://github.com/nim-lang/Nim/pull/14721

### CODE REVIEW

**Comment by Araq** [comment]
Thanks! That looks like a workaround though?

**Comment by timotheecour** [comment]
I think it's a proper fix.

when you extract a routine body B and append a statement S, you need to be aware of the fact that B could contain a `defer` top-level statement. This was the bug, because S was not supposed to run before `B` as it would end iteration (B contained: `complete(future)` in the `async` template/macro expansion, which would set `finished` to the Future)

```nim
proc fn()=
  b1()
  defer:
    b3()
  b2()
=>
proc fnProcessed()=
  b1()
  defer:
    b3()
  b2()
  b4() # <= this will run before `b3`
```

putting the proc body under a block avoids this.

## unrelated to this PR: proposal: `deferScoped`
EDIT: continuing discussion on this in https://github.com/nim-lang/RFCs/issues/236#issuecomment-646855314
regarding https://github.com/nim-lang/RFCs/issues/236, we can introduce
```nim
template deferScoped(cleanup, body) =
  try: body
  finally: cleanup
```
with benefits:
* simpler than `transaction`
* simpler than `defer` since it's a library code (just 1 template)
* like `defer`, brings the "cleanup" close to the code that needs cleaning up
* avoids issues like this there is no scope escaping (but on the flip side, it might be actually a desirable property to extend to end of scope where defer is called)

and drawback:
* introduces a scope

## example
```nim
let a = open("foo.db")
deferScoped: close(a)
do:
  # long rest of code
  use(a)
```

which syntactically improves a bit over try/finally because the cleanup code immediately follows the code that needs cleaning up

File: tests/async/tasync_misc.nim [hunk 1]
```diff
@@ -1,8 +1,3 @@
-discard """
-  exitcode: 0
-  output: "ok"
-"""
-
```
Comment (dom96) [inline tests/async/tasync_misc.nim:5]: Why remove this? I'd rather keep this in case it's relevant to some of the existing tests.

**Review by dom96 — COMMENTED** [review]
I wish you would have explained this in terms of the original test case, would make the explanation easier to read.

**Comment by Araq** [comment]
@dom96 The explanation is very clear, please read it once again when you're not stressed.

File: tests/async/tasync_misc.nim (hunk 1 above)
Comment (timotheecour) [reply tests/async/tasync_misc.nim:5]: it serves no purpose; simpler is better.

**Pull request merged**